workers/unity/Packages/com.improbable.gdk.gameobjectcreation/GameObjectInitializationSystem.cs

```diff
@@ -62,53 +63,46 @@ protected override void OnUpdate()
             for (var i = 0; i < addedEntitiesData.Length; i++)
             {
                 var entity = addedEntitiesData.Entities[i];
-                var spatialEntityId = addedEntitiesData.SpatialEntityIds[i].EntityId;
-
-                var gameObject = gameObjectCreator.OnEntityCreated(
-                    new SpatialOSEntity(entity, EntityManager));
+                var spatialEntityId = addedEntitiesData.SpatialEntityIds[i];
+                var gameObject = gameObjectCreator.OnEntityCreated(new SpatialOSEntity(entity, EntityManager));
 
                 if (gameObject == null)
                 {
                     continue;
                 }
 
-                LinkerSystem.Linker.LinkGameObjectToEntity(gameObject, entity, spatialEntityId, PostUpdateCommands, viewCommandBuffer);
-                entityGameObjectCache.Add(entity, gameObject);
+                linkerSystem.Linker.LinkGameObjectToEntity(gameObject, entity, spatialEntityId.EntityId, viewCommandBuffer);
+                PostUpdateCommands.AddComponent(entity, new HasGameObjectSystemState { EntityId = spatialEntityId });
+                entityToGameObjects.Add(entity, gameObject);
             }
 
             for (var i = 0; i < removedEntitiesData.Length; i++)
             {
                 var entity = removedEntitiesData.Entities[i];
-
-                if (!entityGameObjectCache.TryGetValue(entity, out var gameObject))
-                {
-                    // Entity without linked GameObject removed
-                    gameObjectCreator.OnEntityRemoved(new SpatialOSEntity(entity, EntityManager), null);
-                }
-                else
-                {
-                    entityGameObjectCache.Remove(entity);
-                    LinkerSystem.Linker.UnlinkGameObjectFromEntity(gameObject, entity, true, PostUpdateCommands);
-                    gameObjectCreator.OnEntityRemoved(new SpatialOSEntity(entity, EntityManager), gameObject);
-                }
+                var spatialEntityId = EntityManager.GetComponentData<HasGameObjectSystemState>(entity).EntityId;
+                var gameObject = entityToGameObjects[entity];
```

Comment:
I am guessing that there is no chance of `entityToGameObjects` to not contain a key for the `entity` here because `RemovedEntitiesData` filters via `HasGameObjectSystemState`, but this changes behaviour slightly, the users may still want the `onentityremoved` to be called for entities which do not have corresponding gameobjects.

We can handle that situation in another ticket though, as it is probably not the usual flow.

Reply:
We do want to call it even when there is no gameobject, we will need to predicate this on something different.

Reply:
Why would we want it to be called for entities that have no game objects, if that system is all about creating / deleting gameobjects?

Reply:
this would be interesting, but doesn't sound like it should be in this system. This would most likely be implemented in a separate one

Reply:
Agreed. What they want is us exposing more of the dispatcher and giving users a general signal for entity added/removed. I'm ok with yoloing things together for now but we should revisit this.

Reply:
Matt also said it's reasonable to expect GameObject instantiation requiring cleanup to happen even in cases where they wouldn't want to link. Plus having it symmetric is less of a gotcha. I'm OK with it either way but when I previously asked around I did get the impression that's the functionality people would expect.

Reply:
We definitely want it to be called for everything.
Also no user is going to get confused about it being called on something which has a no op. At worst they get one null ref and work out what is going on and that seems unlikely. The other way round you have a hard to debug issue where you can't work out why you aren't getting a remove entity for something you know was removed.

Reply:
Returning gameobject is a temporary hack to stop users tripping over a current limitation. Eventualyl we will just pass them the linker.

Reply:
However in the cleanup, we only unlink the game object and destroy. If we never create it in the first place, I don't see a reason to have it clean up anything.  This could just cause more troubles...

we should create a separate system, if we want a general callback for added / removed entities

Reply:
That's just that one implementation of the creator. You could also want to return things to the pool, delete non linked entities. Then there is the issue above of how much of a gotcha is it.

Reply:
The alternative is that we make users write two implementions of two different interfaces that receive almost identical information and have to both deal with gameobject lifecycles.

Reply:
Once this is just in a generic gameobject feature module and the linker can support all the stuff it needs, it will just be a generic entity added and removed api that also passes you the worker information and the linker (or you already have the worker information and it just passes you the linker).

It's primary use will then be gamobject lifecycle stuff but it could be whatever. It's just a helper that relies on core and monobehaviours.

Reply:
Another idea was things you can inject that give you worker (rather than entity) callbacks and users bootstrap this somehow on the worker entity. But it is messier and has that bootstrapping problem.

Reply:
They'd have to deal with it in two places if they didn't link something explicitly but it does affect the scene.

Reply:
I don't follow the second paragraph.

Reply:
I'll just revert the change for now. We shouldn't discuss design decisions in a PR. Let's talk more next week when I'm back

Reply:
Eventually it will also be the case that you can link whenever. Or we make something that tell you when things are added and removed and it's up to you to grab a reference to the linker. However those are all variants on this that we can't do right now. But having two sets of static methods seems like it's worse.